### PR TITLE
[Feature](bangpy-ops): Add new operator Celu

### PR DESCRIPTION
**Pull request by SS7D631**
Thanks for your contribution and we appreciate it a lot. 

## 1. Motivation

Please describe your motivation and the goal you want to achieve through this pull request.

## 2. Modification

Please briefly describe what modification is made in this pull request, and indicate where to make the modification.

## 3. Test Report

If you want to know how to do operator testing, you can see [GTest-User-Guide-zh](../docs/GTest-User-Guide-zh.md).

#### 3.1.1 Accuracy Acceptance Standard

For static threshold standard details, see: [MLU-OPS Accuracy Acceptance Standard](../docs/MLU-OPS精度验收标准.md).

- [ ] diff1, diff1 <= 3e-3
- [ ] diff2, diff2 <= 3e-3

#### 3.1.2 Operator Scheme checklist

|     No.      |           Details            |      Check Results       |
|----------------|---------------------------|---------------------|
|        1       |          Supported hardware         | MLU270 <br> MLU290 <br>MLU370|
|        2       |          Job types          |    block <br> U1 <br> U4    |
|        3       |         Layouts            |  NHWC 、NCHW、ARRAY etc    |
|        4       |         Whether multi-dimensions are supported              |                |
|        5       |          Whether element zero is supported             |               |
|        6       |         Data type(half/float)       |         half / float etc           |
|        7      |        Whether there is size limit           |             |

#### 3.1.3 New Feature Test

If you have checked the following items, please tick the relevant box.

- [ ] Data type test
- [ ] Multidimensional tensor test
- [ ] Layout test
- [ ] Different size/integer remainder end segment/alignment misalignment test
- [ ] Zero dimensional tensor test/zero element test
- [ ] stability test
- [ ] Multiple platform test
- [ ] Gen_case module test
- [ ] Nan/INF tests 
- [ ] For memory leak check details, see[GTest-User-Guide-zh](../docs/GTest-User-Guide-zh.md).
- [ ] For code coverage check details, see: [GTest-User-Guide-zh](../docs/GTest-User-Guide-zh.md).
- [ ] For I/O calculation efficiency check details, see: [MLU-OPS Performance Acceptance Criteria](../docs/MLU-OPS性能验收标准.md).

#### 3.1.4 Parameter Check

When a new operator is submitted, the test points are given and the test results are stated.

| Test Point         | Acceptance Criteria | Test Result (Error Message) |
| -------------- | -------- | -------------------- |
| Whether it conforms to the operator restriction | Normal error |                      |
| Whether illegal parameters are passed  | Normal error |                      |

### 3.2 Performance Test

See [MLU-OPS Performance Acceptance Criteria](../docs/MLU-OPS性能验收标准.md) for details.

Platform ：MLU270

|Operator|Mlu_hardware_time(us)|Mlu_interface_time(us)|Mlu_io_efficiency|Mlu_compute_efficiency|Mlu_workwpace_size(Bytes)|Data_type|Shape|
|-----|----|----|----|----|----|------|-----|
|op_name|   |    |     |    |    |    |     |
|op_name|   |    |     |    |    |    |     |

Platform ：MLU290

|Operator|Mlu_hardware_time(us)|Mlu_interface_time(us)|Mlu_io_efficiency|Mlu_compute_efficiency|Mlu_workwpace_size(Bytes)|Data_type|Shape|
|-----|----|----|----|----|----|------|-----|
|op_name|   |    |     |    |    |    |     |
|op_name|   |    |     |    |    |    |     |

Platform：MLU370

|Operator|Mlu_hardware_time(us)|Mlu_interface_time(us)|Mlu_io_efficiency|Mlu_compute_efficiency|Mlu_workwpace_size(Bytes)|Data_type|Shape|
|-----|----|----|----|----|----|------|-----|
|op_name|   |    |     |    |    |    |     |
|op_name|   |    |     |    |    |    |     |

### 3.3 Summary Analysis

Please give a brief overview here, if you want to note and summarize the content.

### CODE REVIEW

File: bangpy-ops/ops/celu/celu.py [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
+# OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+# MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
+# IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
+# CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
+# TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
+# SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+# pylint: disable=missing-docstring, invalid-name, too-many-locals
+"""A multi-platform code link example test for BANGPy TCP."""
```
Comment (SYangDong) [inline bangpy-ops/ops/celu/celu.py:22]: 这个改成对应的算子介绍

File: bangpy-ops/ops/celu/celu.py [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+                    tcp.memcpy(
+                        gram_output[once_loop_start:once_loop_start + calc_size],
+                        nram_buffer_in0[:calc_size])
+
+
+@bangpy.tcp.register_mlu_op(DTYPES, TARGET_LIST, KERNEL_NAME)
+def build_celu(dtype=None, target=None):
+    f = build_module.build(
+        Celu(
+            1,
```
Comment (SYangDong) [inline bangpy-ops/ops/celu/celu.py:116]: 这个地方怎么固定为1呢？

File: bangpy-ops/ops/celu/test_celu.py [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
+# OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+# MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
+# IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
+# CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
+# TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
+# SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+# pylint: disable=missing-docstring, invalid-name, too-many-locals
+"""A multi-platform code link example test for BANGPy TCP."""
```
Comment (SYangDong) [inline bangpy-ops/ops/celu/test_celu.py:22]: 这也地方也是

File: bangpy-ops/ops/celu/test_celu.py [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+
+        (246783,),
+        (246784,),
+        (246785,),
+        (123391,),
+        (123392,),
+        (123393,),
+        (2, 2, 3, 3, 4, 3, 2, 4, 2, 3, 4, 4, 2, 3, 5,),
+    ]
+    for _ in range(append_test_count):
```
Comment (SYangDong) [inline bangpy-ops/ops/celu/test_celu.py:76]: 这个可以写成lambda表达式，可以试试

File: bangpy-ops/ops/celu/test_celu.py [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+            celu_res = output_dev_param.numpy().reshape(primative)
+            return celu_res
+        return celu_inner
+    data_x = np.random.uniform(low=-7.5, high=10, size=shape)
+    f1 = celu_out(alpha)
+    res = f1(data_x.astype(dtype.as_numpy_dtype))
+    torch_value = torch.tensor(data_x)
+    m = torch.nn.CELU(alpha)
+    t_res = m(torch_value)
+    if dtype.name == "float16":
```
Comment (SYangDong) [inline bangpy-ops/ops/celu/test_celu.py:128]: 这个和下面分支的区别是？

Comment (SS7D631) [reply bangpy-ops/ops/celu/test_celu.py:128]: 这里是老代码 原来因为exp接口在half类型下的精度问题 所以改了下断言的误差范围

File: bangpy-ops/ops/celu/celu.py (hunk 2 above)
Comment (SS7D631) [reply bangpy-ops/ops/celu/celu.py:116]: 设置为大于1的编译不通过 设备支持列表里是否有cluster仅为1的设备

Comment (SYangDong) [reply bangpy-ops/ops/celu/celu.py:116]: 那就不需要设置，可以根据target获取

File: bangpy-ops/ops/celu/test_celu.py (hunk 5 above)
Comment (SYangDong) [reply bangpy-ops/ops/celu/test_celu.py:128]: 如果是一样的，那就不用加判断了吧

Comment (SS7D631) [reply bangpy-ops/ops/celu/test_celu.py:128]: 已修改

File: bangpy-ops/ops/celu/test_celu.py (hunk 3 above)
Comment (SS7D631) [reply bangpy-ops/ops/celu/test_celu.py:22]: 已修改

File: bangpy-ops/ops/celu/celu.py (hunk 1 above)
Comment (SS7D631) [reply bangpy-ops/ops/celu/celu.py:22]: 已修改

File: bangpy-ops/ops/celu/celu.py (hunk 2 above)
Comment (SS7D631) [reply bangpy-ops/ops/celu/celu.py:116]: 根据target的str来switch么  目前不清楚每个设备有几个cluster

Comment (ShouyangDong) [reply bangpy-ops/ops/celu/celu.py:116]: 看看其他的例子，
```
tgt = tcp.target()
for i in tcp.thread_binding(0, tgt.cluster_num, thread="blockIdx.x"):
     for j in tcp.thread_binding(0, tgt.core_num, thread="threadIdx.x"):
```
这样就能抓出来，build的时候传入对应的target

Comment (SS7D631) [reply bangpy-ops/ops/celu/celu.py]: 已修改

File: bangpy-ops/ops/celu/test_celu.py (hunk 4 above)
Comment (SS7D631) [reply bangpy-ops/ops/celu/test_celu.py]: 已修改

**Review by ShouyangDong — APPROVED** [review]
